src/core/reference/include/openvino/reference/rms_norm.hpp

```diff
@@ -0,0 +1,62 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+
+#include "openvino/reference/add.hpp"
+#include "openvino/reference/divide.hpp"
+#include "openvino/reference/multiply.hpp"
+#include "openvino/reference/power.hpp"
+#include "openvino/reference/reduce_mean.hpp"
+#include "openvino/reference/sqrt.hpp"
+
+namespace ov {
+namespace reference {
+/**
+ * @brief Reference implementation of RMSNorm operator.
+ *
+ *  Math Formula: (x / Sqrt(ReduceMean(x^2, axes) + eps)) * scale
+ *
+ * @param in           Input pointer to data.
+ * @param axes         Axes for reduce mean calculation
+ * @param out          Output pointer to results.
+ * @param in_shape     Shape of the input Tensor
+ * @param eps          Epsilon for not dividing by zero while normalizing the value
+ * @param scale_shape  Shape of the scale Tensor (optional)
+ * @param scale        Input pointer to scale (optional).
+ *
+ */
+template <class T>
+void rms_norm(const T* in,
+              const AxisSet& axes,
+              T* out,
+              const Shape& in_shape,
+              double eps,
+              const Shape& scale_shape = {},
+              const T* scale = nullptr) {
+    const auto in_elem_count = shape_size(in_shape);
+    std::vector<T> deg_value(in_elem_count, 2);
+    power(in, deg_value.data(), out, in_elem_count);
+
+    const auto reduced_shape = util::reduce_keep_dims(in_shape, axes);
+    const auto reduced_elements_count = shape_size(reduced_shape);
+    std::vector<T> root_mean_square(reduced_elements_count);
+    reduce_mean(out, root_mean_square.data(), in_shape, axes);
+
+    add(root_mean_square.data(),
+        std::vector<T>(reduced_elements_count, static_cast<T>(eps)).data(),
+        root_mean_square.data(),
+        reduced_shape.size());
+    sqrt(root_mean_square.data(), root_mean_square.data(), root_mean_square.size());
+    divide(in, root_mean_square.data(), out, in_shape, reduced_shape, op::AutoBroadcastType::NUMPY, false);
+
+    if (scale) {
+        multiply(out, scale, out, in_shape, scale_shape, op::AutoBroadcastType::NUMPY);
+    }
+}
```

Comment:
```suggestion
template <class T>
void rms_norm(const T* in,
              const AxisSet& axes,
              T* out,
              const Shape& in_shape,
              double eps) {
    const auto in_elem_count = shape_size(in_shape);
    std::vector<T> deg_value(in_elem_count, 2);
    power(in, deg_value.data(), out, in_elem_count);

    const auto reduced_shape = util::reduce_keep_dims(in_shape, axes);
    const auto reduced_elements_count = shape_size(reduced_shape);
    std::vector<T> root_mean_square(reduced_elements_count);
    reduce_mean(out, root_mean_square.data(), in_shape, axes);

    add(root_mean_square.data(),
        std::vector<T>(reduced_elements_count, static_cast<T>(eps)).data(),
        root_mean_square.data(),
        reduced_shape.size());
    sqrt(root_mean_square.data(), root_mean_square.data(), root_mean_square.size());
    divide(in, root_mean_square.data(), out, in_shape, reduced_shape, op::AutoBroadcastType::NUMPY, false);
}

template <class T>
void rms_norm(const T* in,
              const AxisSet& axes,
              T* out,
              const Shape& in_shape,
              double eps,
              const Shape& scale_shape,
              const T* scale) {
      rms_norm(in, axes, out, in_shape, eps);
      multiply(out, scale, out, in_shape, scale_shape, op::AutoBroadcastType::NUMPY);
}          
```

Consider add dedicated functions then `scale_shape` and `scale` data will be mandatory

Reply:
Good idea, done